This patch adds a --slow-down=USECs option that causes the sender to scan
the filelist more slowly, and the generator to scan for deletions more
slowly.  It doesn't do anything to make anyone slow down during the normal
transfer processing, though.

The idea is to lessen rsync's impact on disk I/O.  Unfortunately, there
should really be a way to affect more of rsync's processing, perhaps by
specifying a maximum disk I/O rate (and have that affect a maximum stat()
rate or something like that).

To use this patch, run these commands for a successful build:

    patch -p1 <patches/slow-down.diff
    ./configure                           (optional if already run)
    make

based-on: 54f00c3f89fc147f2f9cba89d26a5bb1d20e783b
diff --git a/flist.c b/flist.c
--- a/flist.c
+++ b/flist.c
@@ -70,6 +70,7 @@ extern int use_safe_inc_flist;
 extern int need_unsorted_flist;
 extern int sender_symlink_iconv;
 extern int unsort_ndx;
+extern unsigned long sleep_asec;
 extern struct stats stats;
 extern char *filesfrom_host;
 
@@ -1666,6 +1667,9 @@ static void send_directory(int f, struct file_list *flist, char *fbuf, int len,
 		}
 
 		send_file_name(f, flist, fbuf, NULL, flags, filter_level);
+		/* Sleep for a bit, to avoid hammering the disk. */
+		if (sleep_asec)
+			usleep(sleep_asec);
 	}
 
 	fbuf[len] = '\0';
diff --git a/options.c b/options.c
--- a/options.c
+++ b/options.c
@@ -108,6 +108,7 @@ int size_only = 0;
 int daemon_bwlimit = 0;
 int bwlimit = 0;
 int fuzzy_basis = 0;
+unsigned long sleep_asec = 0;
 size_t bwlimit_writemax = 0;
 int ignore_existing = 0;
 int ignore_non_existing = 0;
@@ -424,6 +425,7 @@ void usage(enum logcode F)
   rprintf(F,"     --password-file=FILE    read daemon-access password from FILE\n");
   rprintf(F,"     --list-only             list the files instead of copying them\n");
   rprintf(F,"     --bwlimit=KBPS          limit I/O bandwidth; KBytes per second\n");
+  rprintf(F,"     --slow-down=USECs       sleep N usec while creating the filelist\n");
   rprintf(F,"     --write-batch=FILE      write a batched update to FILE\n");
   rprintf(F,"     --only-write-batch=FILE like --write-batch but w/o updating destination\n");
   rprintf(F,"     --read-batch=FILE       read a batched update from FILE\n");
@@ -604,6 +606,7 @@ static struct poptOption long_options[] = {
   {"itemize-changes", 'i', POPT_ARG_NONE,   0, 'i', 0, 0 },
   {"no-itemize-changes",0, POPT_ARG_VAL,    &itemize_changes, 0, 0, 0 },
   {"no-i",             0,  POPT_ARG_VAL,    &itemize_changes, 0, 0, 0 },
+  {"slow-down",        0,  POPT_ARG_LONG,   &sleep_asec, 0, 0, 0 },
   {"bwlimit",          0,  POPT_ARG_INT,    &bwlimit, 0, 0, 0 },
   {"no-bwlimit",       0,  POPT_ARG_VAL,    &bwlimit, 0, 0, 0 },
   {"backup",          'b', POPT_ARG_VAL,    &make_backups, 1, 0, 0 },
